src/main/java/ru/nms/utils/BallTreeUtils.java

```diff
@@ -0,0 +1,263 @@
+
+            if (PROJECTION_COMPARATOR.compare(projection, median) > 0) {
+                putVectorToFile(rightFile, vector);
+            } else {
+                putVectorToFile(leftFile, vector);
+            }
+        }
+        fillMetaData(leftFile, dimension);
+        fillMetaData(rightFile, dimension);
+    }
+
+    public static void verifyFileFormat(FileChannel fileChannel, String fileName) throws IOException {
+        fileChannel.position(0);
+        ByteBuffer headerBuffer = ByteBuffer.allocate(BALL_TREE_HEADER_SIZE).order(ByteOrder.BIG_ENDIAN);
+        fileChannel.read(headerBuffer);
+        headerBuffer.flip();
+        String headerString = new String(headerBuffer.array(), StandardCharsets.UTF_8);
+        if (!BALL_TREE_HEADER.equals(headerString)) {
+            throw new BallTreeFileFormatMismatchException(fileName, headerString);
+        }
+    }
+
+    public static void setChildNodeNames(RandomAccessFile file, int dimension, long leftChildName, long rightChildName) throws IOException {
+        file.seek(CENTROID_POSITION + getVectorSize(dimension));
+        file.writeLong(leftChildName);
+        file.writeLong(rightChildName);
+    }
+
+    public static int getVectorSize(int dimension) {
+        return dimension * Double.BYTES;
+    }
+
+    public static void setPositionToStartOfVectorData(FileChannel fileChannel, int dimension) throws IOException {
+        fileChannel.position(
+                calculateMetadataSize(dimension)
+        );
+    }
+
+    public static void putVectorToFile(RandomAccessFile file, RealVector vector) throws IOException {
+        double[] values = vector.toArray();
+        for (double val : values) file.writeDouble(val);
+    }
+
+    public static RealVector readVectorFromFile(FileChannel fileChannel, int dimension) throws IOException {
+        double[] arr = readNVectorsFromFile(fileChannel, ByteBuffer.allocate(getVectorSize(dimension)).order(ByteOrder.BIG_ENDIAN));
+        return new ArrayRealVector(arr);
+    }
+
+    /**
+     * Вообще говоря надо было и построение дерева переделать на чтение/запись батчами,
+     * но так как скорость построение дерева нас не особо интересует в данном случае,
+     * то на стала тратить на это время, так как суть оптимизации и так ясна на
+     * экспериментах с поиском по дереву.
+     **/
+    public static List<RealVector> readVectorsFromFile(FileChannel fileChannel, int dimension, int vecAmount) throws IOException {
+        double[] vectorsArray = readNVectorsFromFile(fileChannel, getBuffer(dimension, vecAmount));
+        int finalVecAmount = (vectorsArray.length + dimension - 1) / dimension;
+        return IntStream.range(0, finalVecAmount)
+                .parallel()
+                .mapToObj(i -> Arrays.copyOfRange(vectorsArray, i * dimension, Math.min(vectorsArray.length, (i + 1) * dimension)))
+                .map(array -> (RealVector) new ArrayRealVector(array))
```

Comment:
Если это не даункаст, то лучше указывать, что это генерик, чтобы не путать читателя кода:

```suggestion
                .<RealVector>map(array -> new ArrayRealVector(array))
```